### DISCUSSION TITLE
spacy.load() with custom components ---please ignore. Error was caused by other issue///

### DISCUSSION THREAD

**Discussion by python3Berg** [explosion/spaCy#12042] [Help: Coding & Implementations]
Big dumb question. Do the custom components used during training and passed through --code get retained by trained model? If a custom tokenizer is defined during training and subsequently spacy.load(...) is used to deploy, will default or custom tokenizer be used? I see no reference to the custom component in the model directory, but the model loads without error. 

Thanks

**Comment by ljvmiranda921** [comment]
Hi @python3Berg , 

> Do the custom components used during training and passed through --code get retained by trained model?

Yes these should be retained in the trained model. To sanity check, you need to [refer to your custom components in the registry](https://spacy.io/usage/processing-pipelines#custom-components-simple). Ideally the registered names should be in the final / saved configuration file. 

> If a custom tokenizer is defined during training and subsequently spacy.load(...) is used to deploy, will default or custom tokenizer be used?

This really depends on how you set up your tokenizer. If you're modifying the tokenizer directly (e.g., `spacy.load` and then updating the `tokenizer` attribute`), then it won't show up in the config. The best way to ensure that you're using the custom tokenizer is to register it, and refer to it in the configuration file.

**Comment by python3Berg** [comment]
The custom component is in the model-best -> config.cfg. When I do spacy.load() to use in microservice, there are no errors logged. I do not see any of the code anywhere in the model-best directory so I am curious how this all works. Have the various prefixes and suffixes been incorporated into binary model? No errors is generally positive, but never definitive. I could do an import in the service that launches

Additionally, this question was asked another way in different discussion and answered is somewhat different way...
thomashacker
Jan 5, 2023
Maintainer

Hello,
.....

**After you train your model, no custom code will be built-in unless you use spacy package and add the --code parameter referencing your scripts. Another way is to import the custom code to your script when you use spacy.load on trained directories (e.g. ./model-best).**

**Reply by thomashacker** [reply]
Hey, here's a more detailed explanation of how the custom code is incorporated into the model.

When you specify `--code` when using `spacy train` or `spacy project` the custom code gets imported, and all custom functions get registered to the registry so that they can be found when referenced (e.g. when trying to load the config).

A saved (but not packaged) pipeline does not include any code. It uses the code you specified through references to the registry. So if you load the model from disk, you should import your code first, so those are actually in the registry at load time. (Also, you need to be careful about changing the code between training and load time for that reason.) That doesn't have to be complicated; it's just like this:

```
spacy train --code stuff.py xyz.cfg -o xyz
# same as ...
import stuff
spacy.load("xyz")
```
If you package the pipeline, then the specified files are also copied into the package so that they can be imported later. This is designed to be more convenient, but it is absolutely not necessary to use custom code.

However, there are some cases where you only need the code once during initialization (or training), so you don't need it to load the saved model. An example could be initialization callbacks for modifying the tokenizer.

If using `spacy.load()` works for you without importing the custom code, that might be the reason why.

**Reply by python3Berg** [reply]
Thank you so much for the very thorough explanation. This makes sense now...
